### ISSUE TITLE
Windows staticDirs issue - Storybook not loading assets despite path seemingly correct

### ISSUE THREAD

**Issue by LaurenceCliff** [storybookjs/storybook#17271]
**Describe the bug**
On a Windows machine the staticDirs build with the expected structure:
```
dist/
    storybook/
        assets/
            imgs/
                imgs_sub/
                    image.png
            svgs/
                icon.svg
```

Then when we run npm run start, Storybook starts and appears to be working fine, but none of the static assets can be found. When looking at the network it seems to be requesting the correct paths:

http://localhost:6006/assets/imgs/imgs_sub/image.png
http://localhost:6006/assets/svgs/icon.svg

But it's returning a 404. If I go direct to that path it also doesn't load.

main.js looks like this:

```
module.exports = {
  stories: ['../src/**/*.stories.ts'],
  addons: ['@storybook/addon-actions', '@storybook/addon-links', '@storybook/addon-a11y', '@storybook/addon-knobs'],
  staticDirs: [
    '../src/assets',
    { from: '../projects/icons/svgs', to: './assets/svgs' },
    { from: '../projects/images', to: './assets/imgs' }
  ],
  core: {
    builder: 'webpack5'
  },
  angularOptions: {
    enableIvy: false
  }
};
```

Debugging server-statics.js returns the following:

```
_arg$split = [".\projects\icons\svgs",".\assets\svgs"]
target=".\assets\svgs"
staticDir="./.\projects\icons\svgs"
staticPath="C:\Users\user\IdeaProjects\repo\projects\icons\svgs"
targetDir="./.\assets\svgs"
targetEndpoint="/.\assets\svgs"
```
and for imgs

```
_arg$split = [".\projects\images",".\assets\imgs"]
target=".\assets\imgs"
staticDir="./.\projects\images"
staticPath="C:\Users\user\IdeaProjects\repo\projects\images"
targetDir="./.\assets\imgs"
targetEndpoint="/.\assets\imgs"
```

If I remove the dot from main.js, like so, then the build errors:
```
staticDirs: [
    '../src/assets',
    { from: '../projects/icons/svgs', to: '/assets/svgs' },
    { from: '../projects/images', to: '/assets/imgs' }
],
```
Error returned:
```
info => Copying static files: C:\Users\user\IdeaProjects\repo\src\assets at C:\Users\user\IdeaProjects\repo\dist\storybook\
C:\Users\user\IdeaProjects\repo\node_modules\@storybook\core-server\dist\cjs\utils\server-statics.js:118
    throw new Error((0, _tsDedent.default)((0, _chalk.default)`
          ^

Error: Failed to load static files, no such directory: C:\Users\user\IdeaProjects\repo\projects\icons\svgs:\assets\svgs
```

**System**
```
Environment Info:

  System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  Binaries:
    Node: 16.13.2 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.17 - ~\IdeaProjects\repo\node_modules\.bin\yarn.CMD
    npm: 6.14.15 - ~\IdeaProjects\repo\node_modules\.bin\npm.CMD
  Browsers:
    Chrome: 97.0.4692.71
    Edge: Spartan (44.19041.1266.0), Chromium (97.0.1072.62)
  npmPackages:
    @storybook/addon-a11y: ^6.4.13 => 6.4.13
    @storybook/addon-actions: ^6.4.13 => 6.4.13
    @storybook/addon-knobs: 6.2.9 => 6.2.9
    @storybook/addon-links: ^6.4.13 => 6.4.13
    @storybook/addons: ^6.4.13 => 6.4.13
    @storybook/angular: ^6.4.13 => 6.4.13
    @storybook/builder-webpack5: ^6.4.13 => 6.4.13
    @storybook/manager-webpack5: ^6.4.13 => 6.4.13
```

**Comment by shilman** [comment]
@kroeder can you please look into this? i think you're doing this successfully on your project

**Comment by HIMISOCOOL** [comment]
any updates? its pretty annoying to have to manually import all my assets from static.

**Comment by youngchon** [comment]
I just ran into this recently, `staticDir:[..]` instead of `staticDirs:[...]`

**Comment by youngchon** [comment]
> I just ran into this recently, `staticDir:[..]` instead of `staticDirs:[...]`

I think the documentation on using the `staticDir` in the config needs to be update to omit the `s`.

**Comment by lcliff86** [comment]
> I just ran into this recently, `staticDir:[..]` instead of `staticDirs:[...]`

Hi, I've just tried this with staticDir instead of staticDirs and it doesn't output the assets at all that way so I suspect it's something else that's not right.

**Comment by LaurenceCliff** [comment]
> I just ran into this recently, `staticDir:[..]` instead of `staticDirs:[...]`

Hi, I've just tried this with staticDir instead of staticDirs and it doesn't output the assets at all that way so I suspect it's something else that's not right.

**Comment by HIMISOCOOL** [comment]
> It might sound dumb, but for me, I just realized I am using version 6.2, and 6.2 doesn't have `staticDirs` feature, I have to use `start-storybook -s ./static` instead.

thats fair, Im using 6.4 which has OP's issue though

**Comment by LaurenceCliff** [comment]
Hi @kroeder, just wondering if there's any update on this, is it looking like an incorrect config on my side or is something not right with Storybook? Keen not to have to revert back to an older version. Thanks

**Comment by dutscher** [comment]
the same for me. 
i wanted to remove the deprecation warning `DeprecationWarning: --static-dir CLI flag is deprecated` and want to migrate to the staticDirs in main.js.
At first i tried the simple: 
```
module.exports = {
  ...
  staticDirs: ['../public', '../static'],
};
``` which means all subdirs will flatten. 
but dont want the flatten there i want to use `staticDirs: [{ from: '../my-custom-assets/images', to: '/assets' }];`
and now i got the same @LaurenceCliff : my dirs are not exists, but there existing and working with the simple flatten config.
im also work with windows 10.

cheers

---

```
 C:\...\node_modules\@storybook\core-server\dist\cjs\utils\server-statics.js:128
throw new Error((0, _tsDedent.default)((0, _chalk.default)`
          ^

Error: Failed to load static files, no such directory: C:\...\packages\base\assets:\assets
Make sure this directory exists, or omit the -s (--static-dir) option.
    at parseStaticDir (C:\...\node_modules\@storybook\core-server\dist\cjs\utils\server-statics.js:128:11)
    at async C:\...\node_modules\@storybook\core-server\dist\cjs\utils\copy-all-static-files.js:62:34
```

**Comment by dutscher** [comment]
the colon is a seperator for "src":"destination" ;)

i promise the backslashes are the main problem. 
it should do a simple `const osSafePath = path.replace(/\\/g, '/');`

cheers

**Comment by kyro38** [comment]
Are you sure about that ? 
[Here](https://github.com/storybookjs/storybook/blob/92b23c080d03433765cbc7a60553d036a612a501/lib/core-server/src/utils/server-statics.ts), it seems `staticPath` is only a path.

**Comment by justrhysism** [comment]
This might have something to do with it:

<img width="1246" alt="image" src="https://user-images.githubusercontent.com/1326759/156727666-2563a57a-d753-47af-a5ff-451cc0e7fae9.png">

The regex for the split isn't strict enough.

**Comment by justrhysism** [comment]
👆🏼 So I fixed this and submitted a PR, details are (copied from PR):

---

Noticed that the test *"supports absolute file paths with custom endpoint"* didn't cover the reality for Windows.

The `relativeDir` provided by `getDirectoryFromWorkingDir()` resolves all paths to OS separators (e.g. `\` on Windows`). The result of which would look like `C:\\foo\\bar:\\custom-endpoint` and not `C:\\foo\\bar:/custom-endpoint` which the test was covering; so I added a test to cover the **actual** case.

---

Digging in,  the regex split in `parseStaticDir` didn't account for the **actual** path, as it was ignoring all cases of `:\` to cover the `X:\` scenario, which is what resulted in bug #17271. 

To work around this, I changed the code to look for the *last* colon `:`, and use that to split. However, in case the last colon is actually part of a Windows absolute path (e.g. `C:\`), needed to check a few conditions:

- Is the path Windows absolute? (We don't care if Unix absolute because it won't have a colon)
- Is the path just a single path (i.e. doesn't contain the `:` delimiter for static and target dirs.

Once this was determined, the final piece of the puzzle was ensuring the `target` was converted from `\` to `/`, for which I just split on `path.sep` and joined on `path.posix.sep` to ensure a forward-slash `/`.

**Comment by shilman** [comment]
Whoopee!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.5.0-alpha.47 containing PR #17641 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

**Comment by maximedaoust** [comment]
Sorry to jump in a closed issue, but is this problem also linked to the fact that I am not able to use the following ? 
```
    staticDirs: [
      { from: '../../public', to: 'assets/' }, // working properly
      { from: '../../public', to: '/assets/' }, // not working
    ],
```

**Comment by JoshSchreuder** [comment]
Would be nice to backport this to 6.4.x if it's not too much trouble :)

**Comment by JoaquimCAP** [comment]
I got stuck because of this bug and I can't upgrade to another version...
But I found a workaround that allows me to continue:

_You can also configure a directory (or a list of directories) for searching static content when you are starting Storybook. You can do that with the -s flag._
<https://stackoverflow.com/questions/58267903/serving-static-files-in-storybook-js>

Thanks for your great work and all the best for the Storybook team!

**Comment by femi-ojemuyiwa-aa** [comment]
Still doesn't work

**Comment by D-Thrane** [comment]
Did anyone get this to work? I have a folder in the root called public with a svg folder inside. The svg folder ends up in the root of storybook-static.  I need it to put the svg folder inside a folder called public.

**Comment by Tokimon** [comment]
I have experienced this as well. For me/us it is a problem with copying `favicon.ico`. But it is totally random and only shows up once every 10 builds on local, but every time in our docker build.
![image](https://user-images.githubusercontent.com/367237/207875852-b21597ef-e53c-4436-988b-f7e3a921c7ed.png)

We are using version `6.5.14`

**Comment by D-Thrane** [comment]
I also on 6.5.17

Did add this in main.js but it does nothing.
```
  "staticDir": [
    { from: './public/', to: 'public/' },
  ],
```

**Comment by jndietz** [comment]
Yep - experiencing this issue as well with 6.5.14.

```javascript
// main.js
module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
  ],
  framework: "@storybook/react",
  "core": {
    "builder": "@storybook/builder-webpack5"
  },
  staticDirs: ["../assets"],
};
```

```
info @storybook/react v6.5.14
info 
info => Loading presets
info => Serving static files from ./.\assets at / 👈 assets appear to be loaded in, but I don't see them loading in the browser
info => Loading custom babel config as JS
info => Loading custom babel config
```

It should be loading a Roboto font, but it doesn't load, so the browser is falling back to Times New Roman.

**Comment by Tokimon** [comment]
> I have experienced this as well. For me/us it is a problem with copying `favicon.ico`. But it is totally random and only shows up once every 10 builds on local, but every time in our docker build. ![image](https://user-images.githubusercontent.com/367237/207875852-b21597ef-e53c-4436-988b-f7e3a921c7ed.png)
> 
> We are using version `6.5.14`

So we found a solution in our case, which was a mixture of a couple of things

1. We updated Node and pnpm to latest just to be sure.
2. in our solution we copy the file from the application and from storybook into the same folder (/public), which meant that we had favicon.ico twice and it seems that storybook fails because it tries to write the favicon file to a path that already has a one. This is not clear from the error message, though, but renaming our main favicon to something else, seems to fix the issue.

So in short, in our case it seems to be a problem with files with the same name in the destination folder.

**Comment by ravish-goel** [comment]
This is not working with 6.5

main.js config => staticDirs: ['../public'],

This is the logo (in public folder). Check the storybook log at the bottom. It is able to find the directory.
<img width="960" alt="img1" src="https://user-images.githubusercontent.com/27752740/211288732-2c601784-c929-4e60-902a-acc725efaad3.png">

This is how it is displayed.
<img width="960" alt="img2" src="https://user-images.githubusercontent.com/27752740/211288844-dd15f2ad-845e-4e77-a668-b646795266b4.png">

This is how it comes
<img width="960" alt="img3" src="https://user-images.githubusercontent.com/27752740/211289021-4be9fc8c-d36a-418a-becd-dcef12340a8b.png">

**Comment by Mrblackey** [comment]
I confirm that this is still an issue in 6.5.16. :(

**Comment by SherAtrium** [comment]
The same issue in 6.5.13

**Comment by Mrblackey** [comment]
Update:
Problem solved for me after creating `storybook-static` folder at the root of my project and placing all the assets files there.

**Comment by SeanMcMillan** [comment]
Ran into this. It seems that having '..' in the path breaks somewhere in the path normalization chain.

I had `staticDirs: [{ from: '../src/stories/assets', to: '/assets' }],`, and it failed to serve the directory, but when I switched to `staticDirs: [{ from: './assets', to: '/assets' }],` (and moved the files), it worked.

**Comment by jaandrews** [comment]
I'm running into this as well with version 8.2.6 of storybook, but I can't reorganize the files like SeanMcMillan suggests, as it's tied to a nextjs project that requires the current folder structure. I have a nextjs project that is integrated with storybook. My problem is that they handle the public folder differently.

In nextjs, files in the public directory are treated as if they are in the root of the project, so a font would be loaded with
```scss
   @font-face {
       font-family: "ExampleFont";
       src: url("/fonts/ExampleFont.otf") format("opentype")
   }
```
To get it to load in storybook, I have to include "/public" in the url, even though I've set up a static directory to map the "../public/fonts" to "/fonts"
```scss
   @font-face {
       font-family: "ExampleFont";
       src: url("/public/fonts/ExampleFont.otf") format("opentype")
   }
```
The strange thing is that the path without "public" does work when I visit the site on the front end before it throws the error, but that might be due to it running the "@storybook/nextjs" framework, which might be doing something for the front end that isn't factored into the build process.

**Comment by jaandrews** [comment]
Found a different open issue #19340 that refers specifically to the problem I mentioned above.
